File: state-machine/src/main/java/br/com/leonardoferreira/poc/statemachine/service/OrderService.java
package br.com.leonardoferreira.poc.statemachine.service;

import br.com.leonardoferreira.poc.statemachine.statemachine.OrderStateMachineInterceptor;
import br.com.leonardoferreira.poc.statemachine.domain.Order;
import br.com.leonardoferreira.poc.statemachine.domain.enumeration.OrderEvents;
import br.com.leonardoferreira.poc.statemachine.domain.enumeration.OrderStates;
import br.com.leonardoferreira.poc.statemachine.exception.ResourceNotFound;
import br.com.leonardoferreira.poc.statemachine.repository.OrderRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.Message;
import org.springframework.messaging.support.MessageBuilder;
import org.springframework.statemachine.StateMachine;
import org.springframework.statemachine.config.StateMachineFactory;
import org.springframework.statemachine.support.DefaultStateMachineContext;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.UUID;

@Service
public class OrderService {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private StateMachineFactory<OrderStates, OrderEvents> factory;

    @Autowired
    private OrderStateMachineInterceptor orderStateMachineInterceptor;

    @Transactional
    public Long create(BigDecimal price, String sku) {
        Order order = new Order();

        order.setOrderCode(UUID.randomUUID().toString());
        order.setPrice(price);
        order.setSku(sku);
        order.setState(OrderStates.SUBMITTED);

        orderRepository.save(order);

        return order.getId();
    }

    @Transactional
    public StateMachine<OrderStates, OrderEvents> fulfill(Long orderId) {
        StateMachine<OrderStates, OrderEvents> stateMachine = build(orderId);

        Message<OrderEvents> messageEvent = MessageBuilder.withPayload(OrderEvents.FUFILL)
                .setHeader("orderId", orderId)
                .build();
        stateMachine.sendEvent(messageEvent);

        return stateMachine;
    }

    @Transactional
    public StateMachine<OrderStates, OrderEvents> pay(Long orderId) {
        StateMachine<OrderStates, OrderEvents> stateMachine = build(orderId);

        Message<OrderEvents> messageEvent = MessageBuilder.withPayload(OrderEvents.PAY)
                .setHeader("orderId", orderId)
                .build();
        stateMachine.sendEvent(messageEvent);

        return stateMachine;
    }

    @Transactional(readOnly = true)
    public Order findById(Long id) {
        return orderRepository.findById(id)
                .orElseThrow(ResourceNotFound::new);
    }

    @Transactional
    public void updateStatus(Long id, OrderStates state) {
        Order order = findById(id);
        order.setState(state);
        orderRepository.save(order);
    }

    @Transactional
    public StateMachine<OrderStates, OrderEvents> cancel(Long orderId) {
        StateMachine<OrderStates, OrderEvents> stateMachine = build(orderId);

        Message<OrderEvents> cancelMessage = MessageBuilder.withPayload(OrderEvents.CANCEL)
                .setHeader("orderId", orderId)
                .build();

        stateMachine.sendEvent(cancelMessage);

        return stateMachine;
    }

    private StateMachine<OrderStates, OrderEvents> build(Long orderId) {
        Order order = findById(orderId);

        StateMachine<OrderStates, OrderEvents> stateMachine = factory.getStateMachine(order.getOrderCode());

        stateMachine.stop();
        stateMachine.getStateMachineAccessor()
                .doWithAllRegions(sma -> {
                    sma.addStateMachineInterceptor(orderStateMachineInterceptor);
                    sma.resetStateMachine(new DefaultStateMachineContext<>(order.getState(), null, null, null));
                });
        stateMachine.start();

        return stateMachine;
    }

}

File: spring-events/src/main/kotlin/br/com/leonardoferreira/poc/events/service/OrderService.kt
package br.com.leonardoferreira.poc.events.service

import br.com.leonardoferreira.poc.events.domain.entity.Order
import br.com.leonardoferreira.poc.events.domain.mapper.toOrder
import br.com.leonardoferreira.poc.events.domain.mapper.toOrderResponse
import br.com.leonardoferreira.poc.events.domain.request.OrderRequest
import br.com.leonardoferreira.poc.events.domain.response.OrderResponse
import br.com.leonardoferreira.poc.events.events.OrderEvent
import br.com.leonardoferreira.poc.events.exception.ResourceNotFoundException
import br.com.leonardoferreira.poc.events.repository.OrderRepository
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

interface OrderService {
    fun findById(id: Long): OrderResponse
    fun create(orderRequest: OrderRequest): Long
    fun process(order: Order)
}

@Service
internal class OrderServiceImpl(
    private val orderRepository: OrderRepository,
    private val applicationEventPublisher: ApplicationEventPublisher
) : OrderService {
    @Transactional(readOnly = true)
    override fun findById(id: Long): OrderResponse {
        val order = orderRepository.findByIdOrNull(id)
            ?: throw ResourceNotFoundException()
        return order.toOrderResponse()
    }

    @Transactional
    override fun create(orderRequest: OrderRequest): Long {
        val order = orderRepository.save(orderRequest.toOrder())
        applicationEventPublisher.publishEvent(OrderEvent(order))

        return order.id
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    override fun process(order: Order) {
        order.nextStatus()
        orderRepository.save(order)
    }

}
